LIMIT value can be an expression

**Describe the bug**
`LIMIT` can also be values such as `SELECT ...`, `'1'`, etc


**To Reproduce**
Steps to reproduce the behavior:
1. Example SQL:
```
# SELECT * FROM tmp3 LIMIT (SELECT 2);
 order
-------
     1
     2
(2 rows)
# SELECT * FROM tmp3 LIMIT '2';
 order
-------
     1
     2
(2 rows)
```
2. JSQLParser exceptions:
`SELECT * FROM tmp3 LIMIT (SELECT 2)`:
```
Caused by: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "(" "("
    at line 1, column 26.
```
`SELECT * FROM tmp3 LIMIT '2'`:
```
Caused by: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "\'2\'" <S_CHAR_LITERAL>
    at line 1, column 26.
```

**Expected behavior**
For `LIMIT` to be parsed as an expression

**System**
- Database you are using: Postgres
- Java Version: 11
- JSqlParser version: 3.1

--- a/src/main/java/net/sf/jsqlparser/expression/AllValue.java
+++ b/src/main/java/net/sf/jsqlparser/expression/AllValue.java
@@ -1,0 +1,25 @@
+/*-
+ * #%L
+ * JSQLParser library
+ * %%
+ * Copyright (C) 2004 - 2019 JSQLParser
+ * %%
+ * Dual licensed under GNU LGPL 2.1 or Apache License 2.0
+ * #L%
+ */
+package net.sf.jsqlparser.expression;
+
+import net.sf.jsqlparser.parser.ASTNodeAccessImpl;
+
+public class AllValue extends ASTNodeAccessImpl implements Expression {
+
+    @Override
+    public void accept(ExpressionVisitor expressionVisitor) {
+        expressionVisitor.visit(this);
+    }
+
+    @Override
+    public String toString() {
+        return "ALL";
+    }
+}

--- a/src/main/java/net/sf/jsqlparser/expression/ExpressionVisitor.java
+++ b/src/main/java/net/sf/jsqlparser/expression/ExpressionVisitor.java
@@ -182,4 +182,6 @@ public interface ExpressionVisitor {
     void visit(AllColumns allColumns);
 
     void visit(AllTableColumns allTableColumns);
+
+    void visit(AllValue allValue);
 }

--- a/src/main/java/net/sf/jsqlparser/expression/ExpressionVisitorAdapter.java
+++ b/src/main/java/net/sf/jsqlparser/expression/ExpressionVisitorAdapter.java
@@ -498,6 +498,10 @@ public class ExpressionVisitorAdapter implements ExpressionVisitor, ItemsListVis
     }
 
     @Override
+    public void visit(AllValue allValue) {
+    }
+
+    @Override
     public void visit(SelectExpressionItem selectExpressionItem) {
         selectExpressionItem.getExpression().accept(this);
     }

--- a/src/main/java/net/sf/jsqlparser/expression/LongValue.java
+++ b/src/main/java/net/sf/jsqlparser/expression/LongValue.java
@@ -10,6 +10,8 @@
 package net.sf.jsqlparser.expression;
 
 import java.math.BigInteger;
+import java.util.Objects;
+
 import net.sf.jsqlparser.parser.ASTNodeAccessImpl;
 
 /**
@@ -73,5 +75,22 @@ public class LongValue extends ASTNodeAccessImpl implements Expression {
     public LongValue withStringValue(String stringValue) {
         this.setStringValue(stringValue);
         return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        LongValue longValue = (LongValue) o;
+        return stringValue.equals(longValue.stringValue);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(stringValue);
     }
 }

--- a/src/main/java/net/sf/jsqlparser/statement/select/Limit.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/Limit.java
@@ -9,15 +9,15 @@
  */
 package net.sf.jsqlparser.statement.select;
 
+import net.sf.jsqlparser.expression.AllValue;
 import net.sf.jsqlparser.expression.Expression;
+import net.sf.jsqlparser.expression.NullValue;
 import net.sf.jsqlparser.parser.ASTNodeAccessImpl;
 
 public class Limit extends ASTNodeAccessImpl {
 
     private Expression rowCount;
     private Expression offset;
-    private boolean limitAll;
-    private boolean limitNull = false;
 
     public Expression getOffset() {
         return offset;
@@ -35,37 +35,43 @@ public class Limit extends ASTNodeAccessImpl {
         rowCount = l;
     }
 
+    @Deprecated
     public boolean isLimitAll() {
-        return limitAll;
+        return rowCount instanceof AllValue;
     }
 
+    @Deprecated
     public void setLimitAll(boolean b) {
-        limitAll = b;
+        if (b) {
+            rowCount = new AllValue();
+        }
     }
 
+    @Deprecated
     public boolean isLimitNull() {
-        return limitNull;
+        return rowCount instanceof NullValue;
     }
 
+    @Deprecated
     public void setLimitNull(boolean b) {
-        limitNull = b;
+        if (b) {
+            rowCount = new NullValue();
+        }
     }
 
     @Override
     public String toString() {
         String retVal = " LIMIT ";
-        if (limitNull) {
-            retVal += "NULL";
+
+        if (rowCount instanceof AllValue || rowCount instanceof NullValue) {
+            // no offset allowed
+            retVal += rowCount;
         } else {
-            if (limitAll) {
-                retVal += "ALL";
-            } else {
-                if (null != offset) {
-                    retVal += offset + ", ";
-                }
-                if (null != rowCount) {
-                    retVal += rowCount;
-                }
+            if (null != offset) {
+                retVal += offset + ", ";
+            }
+            if (null != rowCount) {
+                retVal += rowCount;
             }
         }
 
@@ -82,11 +88,13 @@ public class Limit extends ASTNodeAccessImpl {
         return this;
     }
 
+    @Deprecated
     public Limit withLimitAll(boolean limitAll) {
         this.setLimitAll(limitAll);
         return this;
     }
 
+    @Deprecated
     public Limit withLimitNull(boolean limitNull) {
         this.setLimitNull(limitNull);
         return this;

--- a/src/main/java/net/sf/jsqlparser/statement/select/Offset.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/Offset.java
@@ -10,50 +10,35 @@
 package net.sf.jsqlparser.statement.select;
 
 import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.JdbcNamedParameter;
-import net.sf.jsqlparser.expression.JdbcParameter;
 
 public class Offset {
-
-    private long offset;
-    private Expression offsetJdbcParameter = null;
+    private Expression offsetExpression = null;
     private String offsetParam = null;
 
-    public long getOffset() {
-        return offset;
+    public Expression getOffset() {
+        return offsetExpression;
     }
 
     public String getOffsetParam() {
         return offsetParam;
     }
 
-    public void setOffset(long l) {
-        offset = l;
+    public void setOffset(Expression offsetExpression) {
+        this.offsetExpression = offsetExpression;
     }
 
     public void setOffsetParam(String s) {
         offsetParam = s;
     }
 
-    public Expression getOffsetJdbcParameter() {
-        return offsetJdbcParameter;
-    }
-
-    public void setOffsetJdbcParameter(JdbcParameter jdbc) {
-        offsetJdbcParameter = jdbc;
-    }
-    
-    public void setOffsetJdbcParameter(JdbcNamedParameter jdbc) {
-        offsetJdbcParameter = jdbc;
-    }
 
     @Override
     public String toString() {
-        return " OFFSET " + (offsetJdbcParameter!=null ? offsetJdbcParameter.toString() : offset) + (offsetParam != null ? " " + offsetParam : "");
+        return " OFFSET " +  offsetExpression + (offsetParam != null ? " " + offsetParam : "");
     }
 
-    public Offset withOffset(long offset) {
-        this.setOffset(offset);
+    public Offset withOffset(Expression offsetExpression) {
+        this.setOffset(offsetExpression);
         return this;
     }
 
@@ -62,7 +47,7 @@ public class Offset {
         return this;
     }
 
-    public <E extends Expression> E getOffsetJdbcParameter(Class<E> type) {
-        return type.cast(getOffsetJdbcParameter());
+    public <E extends Expression> E getOffset(Class<E> type) {
+        return type.cast(getOffset());
     }
 }

--- a/src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java
@@ -433,6 +433,16 @@ public class PlainSelect extends ASTNodeAccessImpl implements SelectBody {
             if (where != null) {
                 sql.append(" WHERE ").append(where);
             }
+
+            if (limit != null) {
+                sql.append(limit);
+            }
+            if (offset != null) {
+                sql.append(offset);
+            }
+            if (fetch != null) {
+                sql.append(fetch);
+            }
         }
         if (forXmlPath != null) {
             sql.append(" FOR XML PATH(").append(forXmlPath).append(")");

--- a/src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java
+++ b/src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java
@@ -13,54 +13,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 
-import net.sf.jsqlparser.expression.AnalyticExpression;
-import net.sf.jsqlparser.expression.AnyComparisonExpression;
-import net.sf.jsqlparser.expression.ArrayExpression;
-import net.sf.jsqlparser.expression.ArrayConstructor;
-import net.sf.jsqlparser.expression.BinaryExpression;
-import net.sf.jsqlparser.expression.CaseExpression;
-import net.sf.jsqlparser.expression.CastExpression;
-import net.sf.jsqlparser.expression.CollateExpression;
-import net.sf.jsqlparser.expression.ConnectByRootOperator;
-import net.sf.jsqlparser.expression.DateTimeLiteralExpression;
-import net.sf.jsqlparser.expression.DateValue;
-import net.sf.jsqlparser.expression.DoubleValue;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.ExpressionVisitor;
-import net.sf.jsqlparser.expression.ExtractExpression;
-import net.sf.jsqlparser.expression.Function;
-import net.sf.jsqlparser.expression.HexValue;
-import net.sf.jsqlparser.expression.IntervalExpression;
-import net.sf.jsqlparser.expression.JdbcNamedParameter;
-import net.sf.jsqlparser.expression.JdbcParameter;
-import net.sf.jsqlparser.expression.JsonAggregateFunction;
-import net.sf.jsqlparser.expression.JsonExpression;
-import net.sf.jsqlparser.expression.JsonFunction;
-import net.sf.jsqlparser.expression.JsonFunctionExpression;
-import net.sf.jsqlparser.expression.KeepExpression;
-import net.sf.jsqlparser.expression.LongValue;
-import net.sf.jsqlparser.expression.MySQLGroupConcat;
-import net.sf.jsqlparser.expression.NextValExpression;
-import net.sf.jsqlparser.expression.NotExpression;
-import net.sf.jsqlparser.expression.NullValue;
-import net.sf.jsqlparser.expression.NumericBind;
-import net.sf.jsqlparser.expression.OracleHierarchicalExpression;
-import net.sf.jsqlparser.expression.OracleHint;
-import net.sf.jsqlparser.expression.OracleNamedFunctionParameter;
-import net.sf.jsqlparser.expression.Parenthesis;
-import net.sf.jsqlparser.expression.RowConstructor;
-import net.sf.jsqlparser.expression.RowGetExpression;
-import net.sf.jsqlparser.expression.SignedExpression;
-import net.sf.jsqlparser.expression.StringValue;
-import net.sf.jsqlparser.expression.TimeKeyExpression;
-import net.sf.jsqlparser.expression.TimeValue;
-import net.sf.jsqlparser.expression.TimestampValue;
-import net.sf.jsqlparser.expression.TimezoneExpression;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.ValueListExpression;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
 import net.sf.jsqlparser.expression.operators.arithmetic.*;
 import net.sf.jsqlparser.expression.operators.conditional.AndExpression;
 import net.sf.jsqlparser.expression.operators.conditional.OrExpression;
@@ -613,6 +566,11 @@ public class TablesNamesFinder implements SelectVisitor, FromItemVisitor, Expres
     @Override
     public void visit(AllTableColumns allTableColumns) {
     
+    }
+
+    @Override
+    public void visit(AllValue allValue) {
+
     }
 
     @Override

--- a/src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java
+++ b/src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java
@@ -13,55 +13,7 @@ import java.util.Iterator;
 import java.util.List;
 import static java.util.stream.Collectors.joining;
 
-import net.sf.jsqlparser.expression.AnalyticExpression;
-import net.sf.jsqlparser.expression.AnalyticType;
-import net.sf.jsqlparser.expression.AnyComparisonExpression;
-import net.sf.jsqlparser.expression.ArrayExpression;
-import net.sf.jsqlparser.expression.ArrayConstructor;
-import net.sf.jsqlparser.expression.BinaryExpression;
-import net.sf.jsqlparser.expression.CaseExpression;
-import net.sf.jsqlparser.expression.CastExpression;
-import net.sf.jsqlparser.expression.CollateExpression;
-import net.sf.jsqlparser.expression.ConnectByRootOperator;
-import net.sf.jsqlparser.expression.DateTimeLiteralExpression;
-import net.sf.jsqlparser.expression.DateValue;
-import net.sf.jsqlparser.expression.DoubleValue;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.ExpressionVisitor;
-import net.sf.jsqlparser.expression.ExtractExpression;
-import net.sf.jsqlparser.expression.Function;
-import net.sf.jsqlparser.expression.HexValue;
-import net.sf.jsqlparser.expression.IntervalExpression;
-import net.sf.jsqlparser.expression.JdbcNamedParameter;
-import net.sf.jsqlparser.expression.JdbcParameter;
-import net.sf.jsqlparser.expression.JsonAggregateFunction;
-import net.sf.jsqlparser.expression.JsonExpression;
-import net.sf.jsqlparser.expression.JsonFunction;
-import net.sf.jsqlparser.expression.KeepExpression;
-import net.sf.jsqlparser.expression.LongValue;
-import net.sf.jsqlparser.expression.MySQLGroupConcat;
-import net.sf.jsqlparser.expression.NextValExpression;
-import net.sf.jsqlparser.expression.NotExpression;
-import net.sf.jsqlparser.expression.NullValue;
-import net.sf.jsqlparser.expression.NumericBind;
-import net.sf.jsqlparser.expression.OracleHierarchicalExpression;
-import net.sf.jsqlparser.expression.OracleHint;
-import net.sf.jsqlparser.expression.OracleNamedFunctionParameter;
-import net.sf.jsqlparser.expression.Parenthesis;
-import net.sf.jsqlparser.expression.RowConstructor;
-import net.sf.jsqlparser.expression.RowGetExpression;
-import net.sf.jsqlparser.expression.SignedExpression;
-import net.sf.jsqlparser.expression.StringValue;
-import net.sf.jsqlparser.expression.TimeKeyExpression;
-import net.sf.jsqlparser.expression.TimeValue;
-import net.sf.jsqlparser.expression.TimestampValue;
-import net.sf.jsqlparser.expression.TimezoneExpression;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.ValueListExpression;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.WindowElement;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
 import net.sf.jsqlparser.expression.operators.arithmetic.Addition;
 import net.sf.jsqlparser.expression.operators.arithmetic.BitwiseAnd;
 import net.sf.jsqlparser.expression.operators.arithmetic.BitwiseLeftShift;
@@ -1037,5 +989,10 @@ public class ExpressionDeParser extends AbstractDeParser<Expression>
     @Override
     public void visit(AllTableColumns allTableColumns) {
         buffer.append(allTableColumns.toString());
+    }
+
+    @Override
+    public void visit(AllValue allValue) {
+        buffer.append(allValue);
     }
 }

--- a/src/main/java/net/sf/jsqlparser/util/deparser/SelectDeParser.java
+++ b/src/main/java/net/sf/jsqlparser/util/deparser/SelectDeParser.java
@@ -329,12 +329,8 @@ public class SelectDeParser extends AbstractDeParser<PlainSelect>
     public void deparseOffset(Offset offset) {
         // OFFSET offset
         // or OFFSET offset (ROW | ROWS)
-        if (offset.getOffsetJdbcParameter() != null) {
-            buffer.append(" OFFSET ").append(offset.getOffsetJdbcParameter());
-        } else {
-            buffer.append(" OFFSET ");
-            buffer.append(offset.getOffset());
-        }
+        buffer.append(" OFFSET ");
+        buffer.append(offset.getOffset());
         if (offset.getOffsetParam() != null) {
             buffer.append(" ").append(offset.getOffsetParam());
         }

--- a/src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java
+++ b/src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java
@@ -9,56 +9,7 @@
  */
 package net.sf.jsqlparser.util.validation.validator;
 
-import net.sf.jsqlparser.expression.AnalyticExpression;
-import net.sf.jsqlparser.expression.AnyComparisonExpression;
-import net.sf.jsqlparser.expression.ArrayConstructor;
-import net.sf.jsqlparser.expression.ArrayExpression;
-import net.sf.jsqlparser.expression.BinaryExpression;
-import net.sf.jsqlparser.expression.CaseExpression;
-import net.sf.jsqlparser.expression.CastExpression;
-import net.sf.jsqlparser.expression.CollateExpression;
-import net.sf.jsqlparser.expression.ConnectByRootOperator;
-import net.sf.jsqlparser.expression.DateTimeLiteralExpression;
-import net.sf.jsqlparser.expression.DateValue;
-import net.sf.jsqlparser.expression.DoubleValue;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.ExpressionVisitor;
-import net.sf.jsqlparser.expression.ExtractExpression;
-import net.sf.jsqlparser.expression.Function;
-import net.sf.jsqlparser.expression.HexValue;
-import net.sf.jsqlparser.expression.IntervalExpression;
-import net.sf.jsqlparser.expression.JdbcNamedParameter;
-import net.sf.jsqlparser.expression.JdbcParameter;
-import net.sf.jsqlparser.expression.JsonAggregateFunction;
-import net.sf.jsqlparser.expression.JsonExpression;
-import net.sf.jsqlparser.expression.JsonFunction;
-import net.sf.jsqlparser.expression.KeepExpression;
-import net.sf.jsqlparser.expression.LongValue;
-import net.sf.jsqlparser.expression.MySQLGroupConcat;
-import net.sf.jsqlparser.expression.NextValExpression;
-import net.sf.jsqlparser.expression.NotExpression;
-import net.sf.jsqlparser.expression.NullValue;
-import net.sf.jsqlparser.expression.NumericBind;
-import net.sf.jsqlparser.expression.OracleHierarchicalExpression;
-import net.sf.jsqlparser.expression.OracleHint;
-import net.sf.jsqlparser.expression.OracleNamedFunctionParameter;
-import net.sf.jsqlparser.expression.Parenthesis;
-import net.sf.jsqlparser.expression.RowConstructor;
-import net.sf.jsqlparser.expression.RowGetExpression;
-import net.sf.jsqlparser.expression.SignedExpression;
-import net.sf.jsqlparser.expression.StringValue;
-import net.sf.jsqlparser.expression.TimeKeyExpression;
-import net.sf.jsqlparser.expression.TimeValue;
-import net.sf.jsqlparser.expression.TimestampValue;
-import net.sf.jsqlparser.expression.TimezoneExpression;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.ValueListExpression;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.WindowElement;
-import net.sf.jsqlparser.expression.WindowOffset;
-import net.sf.jsqlparser.expression.WindowRange;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
 import net.sf.jsqlparser.expression.operators.arithmetic.Addition;
 import net.sf.jsqlparser.expression.operators.arithmetic.BitwiseAnd;
 import net.sf.jsqlparser.expression.operators.arithmetic.BitwiseLeftShift;
@@ -618,6 +569,11 @@ public class ExpressionValidator extends AbstractValidator<Expression> implement
 
     @Override
     public void visit(AllTableColumns allTableColumns) {
+    }
+
+    @Override
+    public void visit(AllValue allValue) {
+
     }
 
 }

--- a/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt
+++ b/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt
@@ -1857,7 +1857,7 @@ PlainSelect PlainSelect() #PlainSelect:
 
     [LOOKAHEAD(<K_OPTIMIZE>) optimize = OptimizeFor() { plainSelect.setOptimizeFor(optimize); } ]
 
-    [ <K_FOR> <K_XML> <K_PATH> "(" token = <S_CHAR_LITERAL> ")" { plainSelect.setForXmlPath(token.image); } ]
+    [LOOKAHEAD(3) <K_FOR> <K_XML> <K_PATH> "(" token = <S_CHAR_LITERAL> ")" { plainSelect.setForXmlPath(token.image); } ]
 
     {
         plainSelect.setSelectItems(selectItems);
@@ -2722,35 +2722,21 @@ JdbcNamedParameter SimpleJdbcNamedParameter() : {
 Limit LimitWithOffset() #LimitWithOffset:
 {
     Limit limit = new Limit();
-    Token token = null;
-    String name;
-    JdbcParameter jdbc = null;
+    Expression rowCountExpression;
+    Expression offsetExpression;
 }
 {
     (
-            LOOKAHEAD(5)
-                // mysql-> LIMIT offset,row_count
-                <K_LIMIT>
-                     (
-                         token=<S_LONG> { limit.setOffset(new LongValue(token.image)); }
-                         |
-                         jdbc = SimpleJdbcParameter() { limit.setOffset(jdbc); }
-                         |
-                         ":" { limit.setOffset(new JdbcNamedParameter()); } [ LOOKAHEAD(2) name = RelObjectNameExt() { ((JdbcNamedParameter)limit.getOffset()).setName(name); } ]
-
-                     )
-                     ","
-
-                (
-                token=<S_LONG> { limit.setRowCount(new LongValue(token.image)); }
-                |
-                 jdbc = SimpleJdbcParameter() { limit.setRowCount(jdbc); }
-                 |
-                 ":" { limit.setRowCount(new JdbcNamedParameter()); } [ LOOKAHEAD(2) name = RelObjectNameExt() { ((JdbcNamedParameter)limit.getRowCount()).setName(name); } ]
-                )
-            |
-            limit = PlainLimit()
+        LOOKAHEAD(<K_LIMIT> Expression() "," Expression()) (
+            // mysql-> LIMIT offset,row_count
+            <K_LIMIT>
+            offsetExpression=Expression() { limit.setOffset( offsetExpression ); }
+            ","
+            rowCountExpression=Expression() { limit.setRowCount( rowCountExpression ); }
         )
+        |
+        limit = PlainLimit()
+    )
     {
         linkAST(limit,jjtThis);
         return limit;
@@ -2760,23 +2746,17 @@ Limit LimitWithOffset() #LimitWithOffset:
 Limit PlainLimit() #PlainLimit:
 {
     Limit limit = new Limit();
-    Token token = null;
-    JdbcParameter jdbc;
-    String name;
+    Expression rowCountExpression;
 }
 {
     // mysql-postgresql-> LIMIT (row_count | ALL | NULL)
     <K_LIMIT>
      (
-         token=<S_LONG> { limit.setRowCount(new LongValue(token.image)); }
-         |
-         jdbc = SimpleJdbcParameter() { limit.setRowCount(jdbc); }
-         |
-        ":" { limit.setRowCount(new JdbcNamedParameter()); } [ LOOKAHEAD(2) name = RelObjectNameExt() { ((JdbcNamedParameter)limit.getRowCount()).setName(name); } ]
-         |
-         <K_ALL> { limit.setLimitAll(true);}
-         |
-         <K_NULL> { limit.setLimitNull(true); }
+        (
+            LOOKAHEAD(3) "(" rowCountExpression = SubSelect() ")"
+            |
+            rowCountExpression = Expression()
+        ) { limit.setRowCount(rowCountExpression); }
      )
     {
         linkAST(limit,jjtThis);
@@ -2787,19 +2767,16 @@ Limit PlainLimit() #PlainLimit:
 Offset Offset():
 {
     Offset offset = new Offset();
-    Token token = null;
-    JdbcParameter jdbc;
-    JdbcNamedParameter jdbcNamed;
+    Expression offsetExpression;
 }
 {
     (
             // postgresql-> OFFSET offset
             // sqlserver-oracle-> OFFSET offset (ROW | ROWS)
              <K_OFFSET>
-                 (token=<S_LONG> { offset.setOffset(Long.parseLong(token.image)); }
-                 | jdbc = SimpleJdbcParameter() { offset.setOffsetJdbcParameter(jdbc); }
-                 | jdbcNamed = SimpleJdbcNamedParameter() { offset.setOffsetJdbcParameter(jdbcNamed); })
-                 [(<K_ROWS> { offset.setOffsetParam("ROWS"); } | <K_ROW> { offset.setOffsetParam("ROW"); })]
+             offsetExpression=Expression() { offset.setOffset( offsetExpression ); }
+
+             [(<K_ROWS> { offset.setOffsetParam("ROWS"); } | <K_ROW> { offset.setOffsetParam("ROW"); })]
 
         )
     {
@@ -3443,20 +3420,23 @@ Expression AnyComparisonExpression() :
 }
 {
  (<K_ANY> { anyType = AnyType.ANY; } | <K_SOME> { anyType = AnyType.SOME; } | <K_ALL> { anyType = AnyType.ALL; } )
+    "("
 
     // if the next block looks alike an ExpressionList without Brackets, then parse as List
-    ( LOOKAHEAD( "(" <K_VALUES> SimpleExpressionList(false) ")" ) 
-        "(" <K_VALUES>
+    (
+        LOOKAHEAD( <K_VALUES> SimpleExpressionList(false) )
+        (
+            <K_VALUES>
             simpleExpressionList = SimpleExpressionList(false) { anyComparisonExpr=new AnyComparisonExpression(anyType, simpleExpressionList).withUseBracketsForValues(false); }
-        ")"
-
-    // Otherwise parse it as a SubSelect
-        | "(" subSelect = SubSelect() { anyComparisonExpr=new AnyComparisonExpression(anyType, subSelect.withUseBrackets(false)).withUseBracketsForValues(false); } ")"
+        )
+         // Otherwise parse it as a SubSelect
+        | subSelect = SubSelect() { anyComparisonExpr=new AnyComparisonExpression(anyType, subSelect.withUseBrackets(false)).withUseBracketsForValues(false); }
             
     )
-   {
+    ")"
+    {
       return anyComparisonExpr;
-   }
+    }
 }
 
 Expression SimpleExpression():
@@ -3701,7 +3681,9 @@ Expression PrimaryExpression() #PrimaryExpression:
 
         | retval=ConnectByRootOperator()
 
-        | retval=Column()
+        | LOOKAHEAD(2) <K_ALL> { retval = new AllValue(); }
+
+        | LOOKAHEAD(2) retval=Column()
 
         | token=<S_CHAR_LITERAL> { retval = new StringValue(token.image); linkAST(retval,jjtThis); }
 
@@ -3712,6 +3694,7 @@ Expression PrimaryExpression() #PrimaryExpression:
         | "{ts" token=<S_CHAR_LITERAL> "}" { retval = new TimestampValue(token.image); }
 
         | LOOKAHEAD("(" retval=SubSelect() ")") "(" retval=SubSelect() ")"
+
         | (
             (
                 (LOOKAHEAD({getAsBoolean(Feature.allowComplexParsing)}) "(" list = ComplexExpressionList() ")"

--- a/src/test/java/net/sf/jsqlparser/expression/LimitExpressionTest.java
+++ b/src/test/java/net/sf/jsqlparser/expression/LimitExpressionTest.java
@@ -1,0 +1,63 @@
+package net.sf.jsqlparser.expression;
+
+import net.sf.jsqlparser.JSQLParserException;
+import net.sf.jsqlparser.parser.CCJSqlParserUtil;
+import net.sf.jsqlparser.statement.select.PlainSelect;
+import net.sf.jsqlparser.statement.select.Select;
+import net.sf.jsqlparser.test.TestUtils;
+import org.junit.Test;
+import org.junit.jupiter.api.Assertions;
+
+/**
+ *
+ * @author <a href="mailto:andreas@manticore-projects.com">Andreas Reichel</a>
+ */
+
+public class LimitExpressionTest {
+    @Test
+    public void testIssue933() throws JSQLParserException {
+        TestUtils.assertSqlCanBeParsedAndDeparsed(
+                "SELECT * FROM tmp3 LIMIT '2'", true);
+
+        TestUtils.assertSqlCanBeParsedAndDeparsed(
+                "SELECT * FROM tmp3 LIMIT (SELECT 2)", true);
+    }
+
+    @Test
+    public void testIssue1373() throws JSQLParserException {
+        TestUtils.assertSqlCanBeParsedAndDeparsed(
+                "SELECT 1 LIMIT 1+0", true);
+
+    }
+
+    @Test
+    public void testIssue1376() throws JSQLParserException {
+        TestUtils.assertSqlCanBeParsedAndDeparsed(
+                "select 1 offset '0'", true);
+
+    }
+
+    @Test
+    public void testMethods() throws JSQLParserException {
+        String sqlStr = "SELECT * FROM tmp3 LIMIT 5 OFFSET 3";
+        Select select = (Select) CCJSqlParserUtil.parse(sqlStr);
+
+        PlainSelect plainSelect = (PlainSelect) select.getSelectBody();
+
+        LongValue longValue = plainSelect.getLimit().getRowCount(LongValue.class);
+        Assertions.assertNotNull( longValue );
+        Assertions.assertEquals( longValue, longValue);
+        Assertions.assertNotEquals( new AllValue(), longValue);
+        Assertions.assertNotEquals( new NullValue(), longValue);
+
+        Assertions.assertNull( plainSelect.getLimit().getOffset(LongValue.class) );
+        Assertions.assertNotNull( plainSelect.getOffset().getOffset(LongValue.class) );
+
+        sqlStr = "SELECT * FROM tmp3 LIMIT ALL";
+        select = (Select) CCJSqlParserUtil.parse(sqlStr);
+        plainSelect = (PlainSelect) select.getSelectBody();
+
+        AllValue allValue = plainSelect.getLimit().getRowCount( AllValue.class );
+        allValue.accept(new ExpressionVisitorAdapter());
+    }
+}

--- a/src/test/java/net/sf/jsqlparser/statement/select/SelectTest.java
+++ b/src/test/java/net/sf/jsqlparser/statement/select/SelectTest.java
@@ -11,6 +11,8 @@ package net.sf.jsqlparser.statement.select;
 
 import java.nio.charset.Charset;
 import static net.sf.jsqlparser.test.TestUtils.*;
+
+import net.sf.jsqlparser.expression.*;
 import org.apache.commons.io.IOUtils;
 
 import static net.sf.jsqlparser.test.TestUtils.assertSqlCanBeParsedAndDeparsed;
@@ -24,6 +26,7 @@ import org.junit.Before;
 import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
+import org.junit.jupiter.api.Assertions;
 import org.junit.rules.TestName;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
@@ -32,19 +35,6 @@ import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import net.sf.jsqlparser.JSQLParserException;
-import net.sf.jsqlparser.expression.BinaryExpression;
-import net.sf.jsqlparser.expression.DoubleValue;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.Function;
-import net.sf.jsqlparser.expression.IntervalExpression;
-import net.sf.jsqlparser.expression.JdbcNamedParameter;
-import net.sf.jsqlparser.expression.JdbcParameter;
-import net.sf.jsqlparser.expression.LongValue;
-import net.sf.jsqlparser.expression.NotExpression;
-import net.sf.jsqlparser.expression.SignedExpression;
-import net.sf.jsqlparser.expression.StringValue;
-import net.sf.jsqlparser.expression.TimeValue;
-import net.sf.jsqlparser.expression.TimestampValue;
 import net.sf.jsqlparser.expression.operators.arithmetic.Addition;
 import net.sf.jsqlparser.expression.operators.arithmetic.Multiplication;
 import net.sf.jsqlparser.expression.operators.arithmetic.Subtraction;
@@ -333,7 +323,7 @@ public class SelectTest {
 
         assertNull(((PlainSelect) select.getSelectBody()).getLimit());
         assertNotNull(((PlainSelect) select.getSelectBody()).getOffset());
-        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffsetJdbcParameter().toString());
+        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffset().toString());
         assertStatementCanBeDeparsedAs(select, statement);
 
         statement = "(SELECT * FROM mytable WHERE mytable.col = 9 OFFSET ?) UNION "
@@ -383,8 +373,8 @@ public class SelectTest {
         rowCount = ((PlainSelect) select.getSelectBody()).getLimit().getRowCount();
 
         assertNull(offset);
-        assertNull(rowCount);
-        assertEquals(3, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
+        Assertions.assertTrue( rowCount instanceof NullValue);
+        assertEquals(new LongValue(3), ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
         assertFalse(((PlainSelect) select.getSelectBody()).getLimit().isLimitAll());
         assertTrue(((PlainSelect) select.getSelectBody()).getLimit().isLimitNull());
         assertSqlCanBeParsedAndDeparsed(statement);
@@ -394,9 +384,12 @@ public class SelectTest {
         offset = ((PlainSelect) select.getSelectBody()).getLimit().getOffset();
         rowCount = ((PlainSelect) select.getSelectBody()).getLimit().getRowCount();
 
+        System.out.println(rowCount.getClass().getName());
+
         assertNull(offset);
-        assertNull(rowCount);
-        assertEquals(5, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
+        Assertions.assertTrue( rowCount instanceof AllValue);
+
+        assertEquals(new LongValue(5), ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
         assertTrue(((PlainSelect) select.getSelectBody()).getLimit().isLimitAll());
         assertFalse(((PlainSelect) select.getSelectBody()).getLimit().isLimitNull());
         assertSqlCanBeParsedAndDeparsed(statement);
@@ -408,7 +401,7 @@ public class SelectTest {
 
         assertNull(offset);
         assertEquals(0, ((LongValue) rowCount).getValue());
-        assertEquals(3, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
+        assertEquals(new LongValue(3), ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
         assertFalse(((PlainSelect) select.getSelectBody()).getLimit().isLimitAll());
         assertFalse(((PlainSelect) select.getSelectBody()).getLimit().isLimitNull());
         assertSqlCanBeParsedAndDeparsed(statement);
@@ -418,7 +411,7 @@ public class SelectTest {
 
         assertNull(((PlainSelect) select.getSelectBody()).getLimit());
         assertNotNull(((PlainSelect) select.getSelectBody()).getOffset());
-        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffsetJdbcParameter().toString());
+        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffset().toString());
         assertStatementCanBeDeparsedAs(select, statement);
 
         statement = "(SELECT * FROM mytable WHERE mytable.col = 9 OFFSET ?) UNION "
@@ -563,10 +556,9 @@ public class SelectTest {
         assertNotNull(((PlainSelect) select.getSelectBody()).getFetch());
         assertEquals("ROWS", ((PlainSelect) select.getSelectBody()).getFetch().getFetchParam());
         assertFalse(((PlainSelect) select.getSelectBody()).getFetch().isFetchParamFirst());
-        assertNull(((PlainSelect) select.getSelectBody()).getOffset().getOffsetJdbcParameter());
         assertNull(((PlainSelect) select.getSelectBody()).getFetch().getFetchJdbcParameter());
-        assertEquals(3, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
-        assertEquals(5, ((PlainSelect) select.getSelectBody()).getFetch().getRowCount());
+        Assertions.assertEquals("3", ((PlainSelect) select.getSelectBody()).getOffset().getOffset().toString());
+        Assertions.assertEquals(5, ((PlainSelect) select.getSelectBody()).getFetch().getRowCount());
         assertStatementCanBeDeparsedAs(select, statement);
     }
 
@@ -582,7 +574,7 @@ public class SelectTest {
         assertEquals("ROW", ((PlainSelect) select.getSelectBody()).getOffset().getOffsetParam());
         assertEquals("ROW", ((PlainSelect) select.getSelectBody()).getFetch().getFetchParam());
         assertTrue(((PlainSelect) select.getSelectBody()).getFetch().isFetchParamFirst());
-        assertEquals(3, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
+        assertEquals(new LongValue(3), ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
         assertEquals(5, ((PlainSelect) select.getSelectBody()).getFetch().getRowCount());
         assertStatementCanBeDeparsedAs(select, statement);
     }
@@ -597,7 +589,7 @@ public class SelectTest {
         assertNotNull(((PlainSelect) select.getSelectBody()).getOffset());
         assertNull(((PlainSelect) select.getSelectBody()).getFetch());
         assertEquals("ROWS", ((PlainSelect) select.getSelectBody()).getOffset().getOffsetParam());
-        assertEquals(3, ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
+        assertEquals(new LongValue(3), ((PlainSelect) select.getSelectBody()).getOffset().getOffset());
         assertStatementCanBeDeparsedAs(select, statement);
     }
 
@@ -627,7 +619,7 @@ public class SelectTest {
         assertNotNull(((PlainSelect) select.getSelectBody()).getFetch());
         assertEquals("ROWS", ((PlainSelect) select.getSelectBody()).getFetch().getFetchParam());
         assertFalse(((PlainSelect) select.getSelectBody()).getFetch().isFetchParamFirst());
-        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffsetJdbcParameter().toString());
+        assertEquals("?", ((PlainSelect) select.getSelectBody()).getOffset().getOffset().toString());
         assertEquals("?", ((PlainSelect) select.getSelectBody()).getFetch().getFetchJdbcParameter().toString());
         assertStatementCanBeDeparsedAs(select, statement);
     }
@@ -942,7 +934,7 @@ public class SelectTest {
         assertEquals(3, ((LongValue) ((PlainSelect) setList.getSelects().get(2)).getLimit().
                 getRowCount()).getValue());
         assertNull(((PlainSelect) setList.getSelects().get(2)).getLimit().getOffset());
-        assertEquals(4, ((PlainSelect) setList.getSelects().get(2)).getOffset().getOffset());
+        assertEquals(new LongValue(4), ((PlainSelect) setList.getSelects().get(2)).getOffset().getOffset());
 
         // use brakets for toString
         // use standard limit syntax

